### DISCUSSION TITLE
OceanBase Community Monthly Report 2022/7

### DISCUSSION THREAD

**Discussion by longdafeng** [oceanbase/oceanbase#1009] [Announcements]
建立一个 **学的轻松, 玩的开心, 用的放心**的开放社区

- 3.1.4 正式发版, OceanBase数据库针对社区项目应用反馈，优化提升内核易用性与稳定性，服务广大社区用户。
   - 内核能力提升：支持数据文件动态按需分配，支持直方图类型统计信息采集与监控。
   - 稳定性提升：扩展单表大事务能力到2TB，多模OBKV支持历史数据过期清理。
   - 易用性提升：支持删除单条缓存计划，优化日志与错误码展示等，ob-operator支持部署OBProxy，开源最新测试套框架。
-  本月组织了8场活动, 有线下社区小伙伴交流会《cto聊天室》, 也有参加外部大会, 比如《架构师大会》/《稀土开发者大会》《CCF 教育论坛》等等, 下个月 也有9场活动, 2场线下meetup, 1场OceanBase 品牌发布会, 外部大会2场等.

**Comment by longdafeng** [comment]
# Roadmap
OceanBase 社区版发布节奏如下：
每2年一个大版本 release，每3个月一次 feature 版本, 每个月一个 bug fix 版本( bp 版本).

- 大版本发布即为架构发生升级， 版本升级类似 MySQL 5.7 升级到 MySQL 8.0, 需要做数据迁移才能完成升级
- feature 版本即为发布了众多 feature 或大 feature , 本地手动冷升级(本地重启)或者通过 OCP 热升级(不停服务)
- bp 版本即为纯 bug fix 版本, 版本升级直接替换 binary 即可, 可以使用 ODP 升级或使用 OCP 热升级

![image.png](https://cdn.nlark.com/yuque/0/2022/png/106206/1658893729965-5c5fadfe-994a-43e2-adc8-52e65b737560.png#clientId=uef63b41f-4fed-4&crop=0&crop=0&crop=1&crop=1&from=paste&height=744&id=u68b80064&margin=%5Bobject%20Object%5D&name=image.png&originHeight=1340&originWidth=2400&originalType=binary&ratio=1&rotation=0&showTitle=false&size=373871&status=done&style=none&taskId=u9b06cde7-dd06-48a9-92d6-8c4cb7ba1c0&title=&width=1333.3333686546052)

**Comment by longdafeng** [comment]
# 社区之星
### 6月月度之星

- 主讲深入浅出第四期 OceanBase 在基于百亿关系知识图谱的新闻流中的应用的郑荣凯
- 热心、主动在社群答疑的罗呈祥
- 主讲深入浅出第五期之携程 OceanBase 拟真压测系统深度解析台枫
- 积极参与征文活动，文章深受社区用户喜欢的唐盛霖
- 主讲深入浅出第五期之携程 OcenanBase 实践数据库发布系统杨晓军

![社区之星](https://user-images.githubusercontent.com/374091/181188585-f5fd6c3d-557f-4000-8f84-b4158130ba33.jpg)

**Comment by longdafeng** [comment]
## 研发进展
## 内核研发
| 六棱镜 | [数据文件按需申请](https://github.com/oceanbase/oceanbase/issues/875) | done, 3.1.4 release |
| --- | --- | --- |
| 京东 | [DML执行失败时, 快速定位失败的列](https://github.com/oceanbase/oceanbase/issues/817) | done, 3.1.4 release |
| 基石数据 | [监控数据增加直方图类型, 更便捷输出至Prometheus](https://github.com/oceanbase/oceanbase/issues/767) | 3.1.4 release |
| 贵州双龙 | [data_dir参数值不准确](https://github.com/oceanbase/oceanbase/issues/876) | 3.1.4 release |
| 快手 | [用户看到的错误码和日志中的一致](https://github.com/oceanbase/oceanbase/issues/877) | 3.1.4 release |
| 快手 | [删除一条 SQL 请求的 plan_cache](https://github.com/oceanbase/oceanbase/issues/944) | 3.1.4 release |
|  |  |  |
| 贝壳 | [多模数据库（Table、HBase API）需要支持自动过期数据清理](https://github.com/oceanbase/oceanbase/issues/874) | 3.1.4 release |
|  | [support load data local ](https://github.com/oceanbase/oceanbase/issues/941) | 详细设计中, 计划3.1.5 release |
|  | [各租户历史日志备份记录保存到表](https://github.com/oceanbase/oceanbase/issues/896) | 详细设计中, 计划3.1.5 release |
|  | [支持django 连接OceanBase](https://github.com/oceanbase/oceanbase/issues/952) | 详细设计中, 计划3.1.5 release |
|  | [clog按照存储大小绝对值限制使用](https://github.com/oceanbase/oceanbase/issues/990) | coding, 计划3.1.5 release |
|  | [增加单表的列上限(512->4096)](https://github.com/oceanbase/oceanbase/issues/950) | coding, 计划3.1.5 release |
|  | [支持MySQL help command](https://github.com/oceanbase/oceanbase/issues/1006) | coding, 计划3.1.5 release |
| 内部需求 | [obclient支持更多mysql-test case 的语句](https://github.com/oceanbase/oceanbase/issues/879) | 编码即将完成, 3.1.5reelase |
| 携程 | [审计日志gv$sql_audit 数据持久化](https://github.com/oceanbase/oceanbase/issues/873) | 编码中,  |
| 携程&58 | [obproxy的proxyro账号的密码按照集群来设置](https://github.com/oceanbase/oceanbase/issues/878) | 编码中, |
| 钉钉用户 | [建议添加list分区根据范围查询可以使用到分区裁剪的功能](https://github.com/oceanbase/oceanbase/issues/840) | done, 但改动较大, 直接做在4.x 版本中, 需要等待4.1 release |
| 红象云腾 | [support latin1 character set](https://github.com/oceanbase/oceanbase/issues/689) | done, 但重新设计整个字符集框架, 会支持更多的字符集格式, 需要在4.1 后, 才能release |
| 用友 | [打开PS, 更好支持golang](https://github.com/oceanbase/oceanbase/issues/331) | done, 但改动遍布很多模块, 为了稳定性, 只能至延期4.0 发布 |
|  |  |  |
|  |  |  |
|  |  |  |



## 生态工具研发
| **需求来源** | **目标和任务详细描述** | **进展详细描述** |
| --- | --- | --- |
| 自动化测试系统 | ob-operator支持部署OBProxy | done, released in 3.1.4 |
| 携程&58 | [独立configure-server](https://github.com/oceanbase/oceanbase/issues/880) | done, released in 3.1.4 |
| 内部需求 | 优化OMS 社区版安装部署 | coding, released in OCP 3.3.0-bp1  |
|  | [Support migration from TiDB to OceanBase]() | done, will be release in OMS 3.3.1 |
|  |  |  |

**Comment by longdafeng** [comment]
# 推荐文章

# 内容进展
## 本月推荐
[向量化引擎对HTAP的价值与技术思考](https://open.oceanbase.com/blog/10900319)    ----> 5 颗星
[影响 Benchmark 性能的 OceanBase 常见操作和场景](https://open.oceanbase.com/blog/10900318)  --> 4 颗星
[高性能数据访问中间件 OBProxy（三）：问题排查和服务运维](https://open.oceanbase.com/)  --> 4 颗星

## 内容专栏
- [专题：SQL 改写系列](https://open.oceanbase.com/blog/10900289?currentPage=1)，已完成 7 篇
-  [专题：OBProxy 系列](https://open.oceanbase.com/blog/10900290?currentPage=1)，已完成3篇
- [专题：OceanBase 源码解读](https://open.oceanbase.com/blog/10900288?currentPage=1)，已完成 15 篇
- [专题：用户实践案例](https://open.oceanbase.com/blog/10900291?currentPage=1)，已完成10篇
- [专题：OceanBase 存储层代码解读](https://open.oceanbase.com/blog/10900287?currentPage=2)，已完成3篇
- [专题：社区故事](https://open.oceanbase.com/blog/10900297?currentPage=1)，已完成 4 篇

![image](https://user-images.githubusercontent.com/374091/181190000-007b3604-2f0a-4569-a82d-8dda33969a89.png)


## 社区专栏

- [拿下 OBCP 认证后，我总结了这份「考试秘籍」](https://open.oceanbase.com/blog/10900313)
- [2022 ArchSummit 全球架构师峰会 精彩回顾](https://open.oceanbase.com/blog/10900315)
- [Paper Time 回顾｜MB2：为自治数据库建立行为模型](https://open.oceanbase.com/blog/10900317)
- [【OceanBase 社区版文档捉虫活动专场】优秀用户名单公布！](https://open.oceanbase.com/blog/10900320)
- [OceanBase 年度征文大赛 | 获奖名单公布！](https://open.oceanbase.com/blog/10900321)
- [对话 ACE 第一期｜国产数据库未来将如何发展？](https://open.oceanbase.com/blog/10900323)
- [50个名额限量开放！「OceanBase 年度发布会」重磅来袭！](https://open.oceanbase.com/blog/10900316)

**Comment by longdafeng** [comment]
# 活动进展
## 下个月活动
| ![image.png](https://cdn.nlark.com/yuque/0/2022/png/106206/1658907776540-07980dd9-bb46-4bef-9b5f-1fd5bfd1a68a.png#clientId=ua268f8a3-b6b0-4&crop=0&crop=0&crop=1&crop=1&from=paste&height=159&id=uc154f64e&margin=%5Bobject%20Object%5D&name=image.png&originHeight=286&originWidth=518&originalType=binary&ratio=1&rotation=0&showTitle=false&size=191939&status=done&style=none&taskId=u4641d9c6-9623-4ff2-8487-a880f9b10b7&title=&width=287.77778540128566) | [分布式数据库 企业降本增效新引擎](https://open.oceanbase.com/activities/4921869) |
| --- | --- |
| ![image.png](https://cdn.nlark.com/yuque/0/2022/png/106206/1658907789186-3129aabb-5005-4405-b4f8-597fa4b3a7f8.png#clientId=ua268f8a3-b6b0-4&crop=0&crop=0&crop=1&crop=1&from=paste&height=120&id=u75f31537&margin=%5Bobject%20Object%5D&name=image.png&originHeight=216&originWidth=508&originalType=binary&ratio=1&rotation=0&showTitle=false&size=139532&status=done&style=none&taskId=ufc426716-ab3c-4e9e-9e72-f43270673a0&title=&width=282.2222296985581) | [2022 OceanBase 年度发布活会](https://open.oceanbase.com/activities) |
| ![image.png](https://cdn.nlark.com/yuque/0/2022/png/106206/1658907806745-15b51da3-65f8-4e09-97f5-b32f3a1a8822.png#clientId=ua268f8a3-b6b0-4&crop=0&crop=0&crop=1&crop=1&from=paste&height=198&id=u868cd4e5&margin=%5Bobject%20Object%5D&name=image.png&originHeight=356&originWidth=634&originalType=binary&ratio=1&rotation=0&showTitle=false&size=261218&status=done&style=none&taskId=u9189e772-8b66-4169-a919-fd0121e502d&title=&width=352.2222315529249) | [《自研分布式数据库的开源之路》](https://pages.segmentfault.com/openinfra-2022), 在线直播 8月5 号 ~ 6 号 |
| ![image.png](https://cdn.nlark.com/yuque/0/2022/png/106206/1658907817905-ccff0878-81ec-411a-89c4-13d0cb817374.png#clientId=ua268f8a3-b6b0-4&crop=0&crop=0&crop=1&crop=1&from=paste&height=152&id=u11294444&margin=%5Bobject%20Object%5D&name=image.png&originHeight=274&originWidth=632&originalType=binary&ratio=1&rotation=0&showTitle=false&size=203497&status=done&style=none&taskId=u4a8c350c-1247-4a12-aad9-3bc56e3722e&title=&width=351.11112041237936) | [http://chinaosc.ccf.org.cn/](http://chinaosc.ccf.org.cn/) |
| ![image.png](https://cdn.nlark.com/yuque/0/2022/png/106206/1658896032483-50f3807d-8306-469d-be74-c0f4eaac2157.png#clientId=uef63b41f-4fed-4&crop=0&crop=0&crop=1&crop=1&from=paste&height=567&id=u1ab1930d&margin=%5Bobject%20Object%5D&name=image.png&originHeight=1020&originWidth=2700&originalType=binary&ratio=1&rotation=0&showTitle=false&size=1319408&status=done&style=none&taskId=u12cd0823-2fef-435c-9ca5-0b9209e5e9a&title=&width=1500.0000397364308) | [《从0到1打造数据库实战教程》1 ](https://open.oceanbase.com/activities), 8月11日：在线直播 |
| ![image.png](https://cdn.nlark.com/yuque/0/2022/png/106206/1658896034725-4bdaecb8-fbc7-4d84-8d0d-ab7dd217c6ec.png#clientId=uef63b41f-4fed-4&crop=0&crop=0&crop=1&crop=1&from=paste&height=567&id=ub2612843&margin=%5Bobject%20Object%5D&name=image.png&originHeight=1020&originWidth=2700&originalType=binary&ratio=1&rotation=0&showTitle=false&size=1319408&status=done&style=none&taskId=ue0d9737b-2fb3-4982-8398-dc157e1e8b4&title=&width=1500.0000397364308) | [《从0到1打造数据库实战教程》](https://open.oceanbase.com/activities)2,  8月17日：在线直播 |
| ![image.png](https://cdn.nlark.com/yuque/0/2022/png/106206/1658896036742-68481b03-83cc-4b46-94f7-e78383bfbcd1.png#clientId=uef63b41f-4fed-4&crop=0&crop=0&crop=1&crop=1&from=paste&height=567&id=u0a8de24f&margin=%5Bobject%20Object%5D&name=image.png&originHeight=1020&originWidth=2700&originalType=binary&ratio=1&rotation=0&showTitle=false&size=1319408&status=done&style=none&taskId=u18774f48-56f5-4430-9c7b-02752453ab8&title=&width=1500.0000397364308) | [《从0到1打造数据库实战教程》](https://open.oceanbase.com/activities)3, 8月23日：在线直播 |
| ![image.png](https://cdn.nlark.com/yuque/0/2022/png/106206/1658896039483-b98fd36d-526c-4469-ad56-487258581952.png#clientId=uef63b41f-4fed-4&crop=0&crop=0&crop=1&crop=1&from=paste&height=567&id=uc9bfdccb&margin=%5Bobject%20Object%5D&name=image.png&originHeight=1020&originWidth=2700&originalType=binary&ratio=1&rotation=0&showTitle=false&size=1319408&status=done&style=none&taskId=uba342a0e-120f-4d8b-907a-5736b4f7ad1&title=&width=1500.0000397364308) | [《从0到1打造数据库实战教程》](https://open.oceanbase.com/activities)4, 8月30日：在线直播 |

# 

## 本月活动
| ![image.png](https://cdn.nlark.com/yuque/0/2022/png/106206/1658906248149-42de27b2-39c0-4d91-a18c-d654217d0169.png#clientId=u97222623-93ee-4&crop=0&crop=0&crop=1&crop=1&from=paste&height=128&id=ud66145b2&margin=%5Bobject%20Object%5D&name=image.png&originHeight=230&originWidth=410&originalType=binary&ratio=1&rotation=0&showTitle=false&size=142289&status=done&style=none&taskId=u40481a2a-7c84-4c08-8b80-749e1c66f5e&title=&width=227.77778381182839) | [**OceanBase CTO 聊天室的活动，北京的小伙伴们成功面基！**](https://ask.oceanbase.com/t/topic/35600667) |
| --- | --- |
| ![image.png](https://cdn.nlark.com/yuque/0/2022/png/106206/1658896905497-f40f6725-7c56-4339-b4e9-6c8512f4eddc.png#clientId=uef63b41f-4fed-4&crop=0&crop=0&crop=1&crop=1&from=paste&height=162&id=ud8dd53a0&margin=%5Bobject%20Object%5D&name=image.png&originHeight=292&originWidth=516&originalType=binary&ratio=1&rotation=0&showTitle=false&size=160973&status=done&style=none&taskId=uc5235265-9f04-49e4-ad9d-91418579a61&title=&width=286.66667426074014) | [Location Cache 模块浅析及OCP监控、报警详解](https://open.oceanbase.com/activities/4921863) |
| ![image.png](https://cdn.nlark.com/yuque/0/2022/png/106206/1658897183199-7eea675a-6dfa-4e2c-b57b-11d0025b59e3.png#clientId=uef63b41f-4fed-4&crop=0&crop=0&crop=1&crop=1&from=paste&height=159&id=u2f73ce6a&margin=%5Bobject%20Object%5D&name=image.png&originHeight=286&originWidth=514&originalType=binary&ratio=1&rotation=0&showTitle=false&size=167438&status=done&style=none&taskId=u5e1aeead-14fd-4fa0-a529-8f4e8afb974&title=&width=285.5555631201946) | [OceanBase 源码解读：一号表与副本位置信息管理 王震南（花名：镇楠](https://open.oceanbase.com/activities/4921854), [OBProxy 功能模块及特性详解 林萌（花名：致新）](https://open.oceanbase.com/activities/4921854) |
| ![image.png](https://cdn.nlark.com/yuque/0/2022/png/106206/1658897319086-46268fb7-ba8c-47c0-8927-5009838c7cfb.png#clientId=uef63b41f-4fed-4&crop=0&crop=0&crop=1&crop=1&from=paste&height=158&id=u2d876cc4&margin=%5Bobject%20Object%5D&name=image.png&originHeight=284&originWidth=510&originalType=binary&ratio=1&rotation=0&showTitle=false&size=154590&status=done&style=none&taskId=ua0dc3eb6-688f-4cd1-8540-f2b99e45498&title=&width=283.33334083910364) | [数据库OBCP认证全面升级公开课    ](https://open.oceanbase.com/activities/4921849), [教你快速拿下 OBCA OBCP        马顺华](https://open.oceanbase.com/activities/4921849) |
| ![image.png](https://cdn.nlark.com/yuque/0/2022/png/106206/1658896818003-4e27d210-f216-4736-8a8e-a5deb57b7c34.png#clientId=uef63b41f-4fed-4&crop=0&crop=0&crop=1&crop=1&from=paste&height=171&id=u1a5f8102&margin=%5Bobject%20Object%5D&name=image.png&originHeight=308&originWidth=526&originalType=binary&ratio=1&rotation=0&showTitle=false&size=109161&status=done&style=none&taskId=u6469bed2-e9fb-4b6b-9685-cd01f91d09e&title=&width=292.22222996346767) | [《分布式数据库的全链路高可用解决方案》](https://open.oceanbase.com/activities/4921859) |
| ![image.png](https://cdn.nlark.com/yuque/0/2022/png/106206/1658896620436-e237e8c5-770e-4751-9c5e-5da995a8a834.png#clientId=uef63b41f-4fed-4&crop=0&crop=0&crop=1&crop=1&from=paste&height=160&id=u3ee5e9b7&margin=%5Bobject%20Object%5D&name=image.png&originHeight=288&originWidth=522&originalType=binary&ratio=1&rotation=0&showTitle=false&size=180412&status=done&style=none&taskId=u122ea62c-0328-41ce-9719-e171639eb51&title=&width=290.00000768237663) | [《携程MySQL 转OceanBase 实践》](https://open.oceanbase.com/activities/4921866),[《OceanBase  读写分离探索与优化》](https://open.oceanbase.com/activities/4921866) |
| ![image.png](https://cdn.nlark.com/yuque/0/2022/png/106206/1658897133135-cd1c715d-20e6-4e55-9b2e-756b7e82644b.png#clientId=uef63b41f-4fed-4&crop=0&crop=0&crop=1&crop=1&from=paste&height=159&id=u21a801e4&margin=%5Bobject%20Object%5D&name=image.png&originHeight=286&originWidth=514&originalType=binary&ratio=1&rotation=0&showTitle=false&size=180376&status=done&style=none&taskId=u16546a96-d48b-439e-89aa-b824433bb26&title=&width=285.5555631201946) | [开放式时空大数据助力智能公交路线规划](https://open.oceanbase.com/activities/4921861) |
| ![image.png](https://cdn.nlark.com/yuque/0/2022/png/106206/1658897532294-612a2d55-ba7f-4db2-9d71-8ffb105bd62f.png#clientId=uef63b41f-4fed-4&crop=0&crop=0&crop=1&crop=1&from=paste&height=218&id=uac1055c2&margin=%5Bobject%20Object%5D&name=image.png&originHeight=392&originWidth=1294&originalType=binary&ratio=1&rotation=0&showTitle=false&size=647327&status=done&style=none&taskId=u082a5eb1-e4cf-4317-92b9-9f4607ce7d9&title=&width=718.8889079329414) | [国产数据库系统自主研发人才培养的瓶颈与对策](https://mp.weixin.qq.com/s/v0E5Ph0AtqLI7XtH_QEWxQ) |

**Comment by longdafeng** [comment]
# 故障排查 SOP 系列
**SOP 系列累积 已完成12 篇内容。**
1、【SOP 系列 13】备份恢复 (编写中)

**Comment by Xjxjy** [comment]
# 客户声音

**精选客户问题**

|    问题描述    |    分析       |
|---------------|---------------|
| Oceanbase 的 multi-paxos 的实现方式， 相比 multi-raft 有哪些优势？                                                                                                                                                                                                                                | multi-paxos 跟 multi-raft 不应该直接相比。他们两个的 multi 不是一个东西。multi-paxos 的 multi 是指一次 proposal 可以提交多个值。multi-raft 的 multi 是指有很多个 raft 组。</br>multi-paxos 相对于 raft 最大的优势是提交的数据，可以乱序，可以有空洞。而 raft 要求必须严格递增，不允许有空洞。这点在性能上 paxos 会有比较大的优势，特别是网络延迟稍微大的时候，比如跨城机房。                                                                                                                                                                                                                                                                                                                                                                                                                                                                      |
| Clog 日志和 Redo 日志有什么区别                                                                                                                                                                                                                                                                 | Redo 日志是 OceanBase 数据库用于宕机恢复以及维护多副本数据一致性的关键组件。Redo 日志是一种物理日志，它记录了数据库对于数据的全部修改历史，具体的说记录的是一次写操作后的结果。从某个持久化的数据版本开始逐条回放 Redo 日志可以还原出数据的最新版本。</br> OceanBase 数据库的 Redo 日志文件包含两种类型：<ul><li>Clog：用于记录 Redo 日志的日志内容，位于 `store/clog` 目录下。</li><li> ilog：用于记录相同分区相同 Log ID 的已经形成多数派日志的 Commit log 的位置信息。位于 store/ilog 目录下。</li></ul> 详细介绍可参考官网文档：https://open.oceanbase.com/docs/observer-cn/V3.1.4/10000000000449455                                                                                                                                                                                                                                                                                                                                                           |
| 一张表可以容纳的最大数据量是多少？32G 内存大概多少呢？有数字吗？                                                                                                                                                                                                                                                  | 表的容量和内存无关，和磁盘有关；目前 OceanBase 的架构是一个分区副本的数据全部在一台 Server 上；</br> 所以如果你的表是分区表，并且每个分区都分布在不同的机器上，那么这张表的理论上限存储空间是 `单机磁盘容量 * 分区数`，当然通常一个机器上包含同一个表的多个分区，并且可能存在转储等占用额外的空间，所以这个理论上限往往达不到；</br> 如果你的表是非分区表，那么这张表的上限就取决于单机磁盘容量。                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        |
| OMS 全量同步原理是直接 select 拉取表中的数据吗？</br> 上游有新的更新，开启增量后，是怎么衔接的？是通过回调拉取时间吗？</br> OMS 应该不是常规的 CDC 处理逻辑，开启增量前先开启增量拉取模块，全量完成后再回放增量拉取？                                                                                                                                                                  | 全量同步原理，可以简单理解为查询后批量插入，当然有优化处理，要分片等等。</br> 对于如何衔接这一问题，会先增量落盘，再开启全量迁移，整个的过程中，全量与增量有重叠部分，基于 at least once 设计，重复执行会幂等，一定是最终一致的。不选择增量迁移，那就相当于一次全量迁移，这种情况，是没有办法最终一致的（假设上游没有停止写，因为全量拿到的数据是快照，无增量相当于无补偿行为，如果上游停写，就另说）。增量拉取 与 全量迁移数据 是重叠的。                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 |
| 每次 OCP 服务器重启后，都要按照启动 ocp_meta 所在的 OBServer 进程-\>启动 OBProxy-\>再 docker start ocp 的顺序手工执行执行一遍相关的命令？                                                                                                                                                                                  | 分两种情况：<ul><li>如果 OCP 和 MetaDB 分开部署到不同的机器，确认 OCP 对应的 MetaDB 的 OceanBase 环境是正常的，只需要执行 docker start ocp 启动 OCP 容器就可以。</li><li> 如果 OCP 和 MetaDB 部署到一起，需要先启动 MetaDB 的 OceanBase 环境和对应的 OBProxy，最后启动 OCP 容器。                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   |
| 一个表的 data_size 与磁盘上实际占用的空间能用 SQL 查出来吗？                                                                                                                                                                                                                                             | 登录至数据库中，查询表 `__all_virtual_meta_table`，data_size 列显示的是压缩之前的，即原始的字符串大小；required_size 列显示的是实际磁盘占用大小。                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    |
| 1、表的scheme元数据管理在哪里，怎么做到在线 DDL？</br>2、虚拟表是不是物理表，还是只是内存中的数据呢？                                                                                                                                                                                                                           | 1、表的 schema 存在内部表中，可以参考：OceanBase 源码解读（十三）：一号表 https://zhuanlan.zhihu.com/p/503059729 </br>2、虚拟表是内存中的数据，并没有直接的物理存储。                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           |
| ODC 升级问题：升级执行数据迁移失败，从 3.1.x- 版本升级到 3.2.0+ 版本，内置 admin 账号名称冲突                                                                                                                                                                                                                       | 为避免账号冲突，ODC 启动时会进行预检查，如果存在内置账号名称冲突则需要将之前版本冲突的账号重命名。ODC 会生成一个 rename.properties 文件对系统中的同名 admin 用户进行更名操作。将新的账户名填在冲突名称之后，再重启应用即可解决该问题。</br>解决方案（以用户名 admin 冲突为例） ：</br>1. 在安装目录下的 ../odc/static/tmp/rename.properties 中编辑 rename.properties，格式为 admin=用户重命名。</br> 2. 用户名修改完成后保存，再次尝试启动 ODC。                                                                                                                                                                                                                                                                                                                                                                                                                                            |
| 在不使用 `obd cluster restart` 重启整个集群的前提下，如何使用 OBD 启动已经挂掉的单个节点呢？                                                                                                                                                                                                                         | <ol><li> 先使用 obd upgrade 命令将 OBD 升级到最新版本，运行 obd cluster start --help ，可以看到有 -s 选项。</li><li> 没有连互联网时需要手动更新一下 OBD 的 RPM 包。</li><li> 启动指定的节点</li><ol><li>使用 -s 选项指定节点名：-s SERVERS 或 --servers=SERVERS，节点名是 OBD yaml 文件里 oceanbase-ce 下 servers 里对应的 name 。</li><li>执行 obd cluster start deploy_name -s SERVERS  </li></ol></li></ol>                                                                                                                                                                                                                                                                                                                                                                                                                                                        |
| OBD 重启 OBServer，一直显示 `wait for observer init`                                                                                                                                                                                                                                       | 【问题原因】：资源不足，clog 存储不够 。</br>【解决方案】：扩存储，另外建议把 data 和 clog 分盘部署。创建软连接，将 `redo_dir: /data/clog` 对应的放在新的盘上，然后把 `/data/clog` 做个软连接指向新的盘的路径。</br> clog、ilog、slog 都要做软连接，可以在老的路径下创建一个测试文件，查看是否在新的路径下，并且权限正确。clog 的大小，一般建议是内存的 4 倍，可以适当分配多一些，避免出现 clog 盘满了集群异常。                                                                                                                                                                                                                                                                                                                                                                                                                                                                         |
| 传统数据库在进行参数化时一般是对语法树进行参数化，然后使用参数化后的语法树作为键值在计划缓存中获取计划。</br>对语法树进行参数化，怎么理解？                                                                                                                                                                                                               | 比如 `select * from test where c1=1 → select * from test where c1=?` ，即将 SQL 语句中的常量参数化。</br>然后这个问题应该来想了解快速参数化？其实一些场景下只需要一个参数化后的 SQL 去处理，OceanBase 基于 lex/yacc 实现针对整个 SQL 层的 token 的状态转换开销大，也就是说传统的 parser 需要识别出所有的 token, 但 fast parser 只需要识别出 SQL 中应该被忽略的部分、应该被原样保留的部分、以及应该被参数化的常量部分。                                                                                                                                                                                                                                                                                                                                                                                                                                                |
| obloader 导入表大小写问题：</br>MySQL 模式，社区版；</br>建了一个大写的表比如 T_PRODUCT，用obloader 导入 t_product.csv，--table 'T_PRODUCT' 导入失败，报 Load csv file 失败；报错代码行 `DBUtils.calculatePartitionId(Line:1670)`。但是把 T_PRODUCT 表删了，改成小写名字 t_product，导入同一个 csv，--table 't_product' 就成功导入了。 | OceanBase MySQL Mode 和 MySQL Community Server 行为是一致的，默认创建的表名是小写的。比如 `MySQL \> create table TEST (c1 INT);` -- 默认创建出来的表名是小写的，等同于 `create table test (c1 INT);` </br>可以修改变量 lower_case_table_names 的值，打开大小写敏感模式 `show variables like 'lower_case_table_names';` ，该变量打开以后，create table test 与 create table TEST 是不同的表名。</br> 提示：obloader/obdumper 兼容数据库的行为，默认是小写的，数据文件名也要求是小写的。如果要考虑大写的文件名，可以指定 --table '\[TEST]' 表示数据库中的表是 TEST 而不是 test，同时磁盘的文件也要求是 TEST.csv 而不是 test.csv。当数据库打开大小写敏感模式，这种设计是为了防止数据库中同时存在 TEST，test 两张表。obloader 可以正常识别文件和表结构。 --table 'test' 与 --table 'TEST' 在 MySQL 模式下都表示小写的 test; --table 'test' 与 --table '\[TEST]' 在 MySQL 模式下分别表示 test 和 TEST； 中括号用来显式指定大写的表名，否则 obloader 像 MySQL一样默认把表名转换成小写的。 |
| OCP 服务器没有业务数据，但是对硬件配置要求过高了？                                                                                                                                                                                                                                                        | 文档里提到的配置是生产环境的经验值，具体所需和 OCP 管理集群的规模有关系，OCP 设计上是用来管理多个 OceanBase 集群的，所以资源使用上会占用比较多，如果仅用来测试或者管理很小规模的集群，也可以在资源上进行调整，如果觉得单独搭建一个集群比较费资源，可以直接分给 OCP 两个租户。</br> 如果用了比较少的资源，需要结合实际的场景进行测试，关注租户的转储情况，CPU 负载， OCP gc 的情况。 另外，尽量用 OCP 来管理更多的集群，这样平均下来资源占用会相对没那么多，如果一开始用了小规格部署，后面 OceanBase 集群规模增加时，也要做对应的调整。                                                                                                                                                                                                                                                                                                                                                                                                                       |
| 【问题描述】</br>使用 OCP 接管 ARM 集群报 t-oceanbase-ocp-agent,version=3.3.0,architecture=aarch64不存在</br>【问题现象及影响】</br>配置完机型和凭据后点击提交后报上述错误，手动上传 obagent-1.1.2-9.el7.aarch64.rpm 软件包后再次提交报相同错误，X86 版 OCP 是否支持接管 ARM OceanBase 集群                                                                                                  | 目前 X86 版的 OCP 是不支持接管 ARM 的 OceanBase 集群的，接管 ARM 架构下OceanBase 集群需要 ARM 的 OCP 镜像。预计下个版本正式提供下载 ARM 版镜像，且在 ARM 主机上部署 x86 的 OCP 镜像是不行的，x86 的镜像只能在 x86 主机上部署。                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     |
| 1、我在命令行里创建的资源单元 unit_odc，在租户-Unit规格管理里无显示(已经使用该资源单元成功创建了 ODC 所使用的租户)</br> 2、租户-新建租户，Unit 规格里很多规格不显示，包括 “Unit规格管理” 里系统默认规格也无法显示 </br> 3、为什么这里创建租户会跳过 “创建资源单元-创建资源池-创建租户” 流程里创建资源池这一步，是自动创建了资源池吗？                                                                                                        | 1 和 2 对小于 5G 的规格，OCP 的规格页面做了过滤（不展示），OCP 会自动创建资源池。连接 OCP 的 MetaDB 修改参数可解决创建租户时的限制：</br> select \* from config_properties where \`key\` like '%small%'UPDATE config_properties SET value='true' WHERE \`key\` = 'ocp.operation.ob.tenant.allow-small-unit'</br> 3、会自动创建资源池                                                                                                                                                                                                                                                                                                                                                                                                                                                           |
| OceanBase 会考虑实现表锁吗？</br>比如在执行 DML 事务的期间，在另一个 session 中执行 truncate。会出现 DML 事务执行中，表被清空，提交后有数据丢失现象发生。                                                                                                                                                                                      | 在 OceanBase 3.x 版本中，truncate 相当于先删除表，再新建表，目前看到的现象是符合预期的；为避免出现这个现象，可以把 truncate 改成 delete。</br> OceanBase 4.x 版本会重构 truncate，解决上面的问题；4.0 版本实现了表锁，truncate table 会一直等到另一个 session 的事务提交之后再结束。                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                |
| OMS 从 OceanBase 数据库反向增量同步到 MySQL，是什么原因导致分区键不支持更新操作？</br>OMS 官方文档说是不支持的，自己测试也是确实当 OceanBase 端更新了分区键的时候，源端 MySQL 不会去同步。想咨询一下是什么原因导致这种限制，有没有好的解决方法？                                                                                                                                                               | 对于含唯一索引的分区表，如果发生 row movement (即更新分区键的值)，在 OceanBase 会在新的分区产生一个 insert 的日志，在旧的分区产生 delete 日志，两条日志对应同一个事务。对于 OceanBase 而言没有什么影响，但因为 OceanBase 分区级日志流，因此对于数据链路来说，insert 和 delete 输出顺序是不确定的，如果先输出 insert，再输出 delete，由于唯一索引的存在，可能会将数据 delete 掉，导致目标端和源端数据不一致的问题。</br> 为了解决这个问题，OceanBase 在 3.2 及之后的版本中引入了 seq_no 标识数据的操作顺序，保证上述场景中的 delete 操作对应的 seq_no 小于 insert 操作的 seq_no，obcdc 按 seq_no 排序输出行数据，由此保证数据链路的正确性。注意是在 3.2.x 版本之后才有的能力，在 3.2.x 之前及社区版，是没有这个保证的。 用户可以在 OceanBase 数据库中设置禁止 row movement，如果发生 row movement，sql 会报错，SQL 可以参考：`alter table xxx disable row movement;`                                                                                                                                                                                             |
